_posts/2021-11-04-prefix-sum-portable.md

```diff
@@ -0,0 +1,147 @@
+
+There is a [WebGPU implemenation of decoupled look-back] by Reese Levine, but at this point it's not expected to pass. Hopefully it will be useful to shake out bugs (it would pass if the barrier scope were upgraded), and also to point the way for extensions to WebGPU that would allow this algorithm to run.
+
+### Shader translation
+
+Another concern I have about WebGPU is that it is difficult to maintain shaders as single authoritative source files and translate them automatically into WGSL. For all other shader languages, we've arrived at a reasonable solution involving GLSL source and translation using spirv-tools to HLSL and Metal (enabling translation to intermediate representation as well). That won't work for WGSL, and, not surprisingly, the problem is atomics.
+
+The fundamental issue is that [atomics][Atomics proposal] in WGSL are strictly typed, following the precedent of Metal, which in turn follows C++. By contrast, in HLSL and GLSL, atomics are operations that can work on ordinary memory locations. Arguably, that's not ideal, but it is how shaders are written today. That doesn't create a portability problem in practice, because Metal also supports unsafe C++ casting operations, and this pattern is widely used (by spirv-cross and perhaps ironically also [Tint]) to compile the standard atomics of shader languages into Metal.
+
+WGSL can't expose unsafe casting, so that's not a viable solution. Unfortunately, this means that translation into WGSL is not a straightforward problem, but in some cases might require rearchitecting the shader and the pipeline that contains it. A very common pattern (used in piet-gpu) is to treat memory as a big array of 32 bit integers (RWByteAddressBuffer), and store all the types in there. It's primitive, but low level shader languages aren't really expressive enough to represent richer types properly, especially if you want to pack them efficiently or have dynamic layout. That doesn't work when some of the types have atomic fields; they have to be in a separate space recognized by the compiler.
+
+Maintaining source in WGSL and translating to other shader languages would be viable, but that would also limit the code to using only least common denominator features. Perhaps WGSL will grow a rich set of optional extensions, in which case it might make sense. In any case, at least for my needs, it is a fairly major point of friction.
+
+It's worth noting that the Vulkan memory model faced a similar choice when they were defining the semantics of atomics for Vulkan, and came to a different decision than WGSL. As is clear from [their blog post,][Atomic Operations vs Atomic Objects] they recognize the importance of working with existing shaders and shader languages, even if it isn't as aesthetically pleasing as a strictly typed approach.
```

Comment:
Yes, WGSL/WebGPU targets the intersection of features in underlying APIs, at least for the first version.  Vulkan targets the union of features of underlying implementations.